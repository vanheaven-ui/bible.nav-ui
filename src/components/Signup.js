import { useState } from 'react';
import { Link, useHistory } from 'react-router-dom';
import '../styles/signup.css';

const Signup = () => {
  const [username, setUsername] = useState('');
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [passwordAgain, setPasswordAgain] = useState('');

  // state variables to manage password type
  const [passwordType, setPasswordType] = useState('password');

  const hist = useHistory();

  const [registering, setRegistering] = useState(false);

  const signupParams = {
    username,
    email,
    password,
    password_confirmation: passwordAgain,
  };

  const toggle = e => {
    if (e.target.checked) {
      setPasswordType('text');
    } else if (!e.target.checked) {
      setPasswordType('password');
    }
  };

  const handleSubmit = e => {
    e.preventDefault();
    setRegistering(true);
    fetch('https://biblenav-api.herokuapp.com/api/v1/signup', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(signupParams),
    })
      .then(res => {
        setRegistering(false);
        if (res.ok) {
          hist.push('/login');
        }
      });
  };

  return (
    <section className="signup">
      <h3 className="h4">Register here to have Bible.nav priviledges</h3>
      <form onSubmit={e => handleSubmit(e)}>
        <div className="form-group">
          <input
            type="text"
            placeholder="Username"
            value={username}
            onChange={e => setUsername(e.target.value)}
          />
        </div>
        <div className="form-group">
          <input
            type="text"
            placeholder="Email"
            value={email}
            onChange={e => setEmail(e.target.value)}
          />
        </div>
        <div className="form-group">
          <input
            type={passwordType}
            placeholder="Password"
            value={password}
            onChange={e => setPassword(e.target.value)}
          />
          <p className="show-hide">
            <small>show password</small>
            {' '}
            <input type="checkbox" id="show-password" onClick={e => toggle(e)} />
          </p>
        </div>
        <div className="form-group">
          <input
            type={passwordType}
            placeholder="Type password again"
            value={passwordAgain}
            onChange={e => setPasswordAgain(e.target.value)}
          />
        </div>
        <div className="actions">
          { !registering && <button type="submit" className="block-btn">Register</button> }
          { registering && <button type="button" disabled className="block-btn">Registering...</button> }
        </div>
      </form>
      <div className="other-action">
        Already registered?
        {' '}
        <Link to="/login">Login Here</Link>
      </div>
    </section>
  );
};

export default Signup;
