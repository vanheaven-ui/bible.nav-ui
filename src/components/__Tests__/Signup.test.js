import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { render, screen, fireEvent } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import rootDiv from '../../testHelper';
import '@testing-library/jest-dom';
import Signup from '../Signup';

it('renders without crashing', () => {
  ReactDOM.render(<Router><Signup /></Router>, rootDiv());
});

it('does not change unexpectedly', () => {
  const tree = renderer.create(
    <Router><Signup /></Router>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders the correct wrapper element', () => {
  const { container } = render(
    <Router><Signup /></Router>,
  );
  expect(container.firstChild.classList.contains('signup')).toBe(true);
});

it('displays loading message on login click', () => {
  render(
    <Router><Signup /></Router>,
  );
  fireEvent.click(screen.getByText('Register'));
  expect(screen.getByRole('button')).toHaveAttribute('disabled');
  expect(screen.getByRole('button')).toHaveTextContent('Registering');
});
