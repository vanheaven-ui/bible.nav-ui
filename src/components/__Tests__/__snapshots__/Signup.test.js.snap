// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not change unexpectedly 1`] = `
<section
  className="signup"
>
  <h3
    className="h4"
  >
    Register here to have Bible.nav priviledges
  </h3>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        onChange={[Function]}
        placeholder="Username"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <input
        onChange={[Function]}
        placeholder="Email"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <input
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
      <p
        className="show-hide"
      >
        <small>
          show password
        </small>
         
        <input
          id="show-password"
          onClick={[Function]}
          type="checkbox"
        />
      </p>
    </div>
    <div
      className="form-group"
    >
      <input
        onChange={[Function]}
        placeholder="Type password again"
        type="password"
        value=""
      />
    </div>
    <div
      className="actions"
    >
      <button
        className="block-btn"
        type="submit"
      >
        Register
      </button>
    </div>
  </form>
  <div
    className="other-action"
  >
    Already registered?
     
    <a
      href="/login"
      onClick={[Function]}
    >
      Login Here
    </a>
  </div>
</section>
`;
